gh-119670: Add 'always' keyword argument to `shlex.quote`

jb2170 opened the pull request:
Addresses #119670 

<!-- gh-issue-number: gh-119670 -->
* Issue: gh-119670
<!-- /gh-issue-number -->


<!-- readthedocs-preview cpython-previews start -->
----
📚 Documentation preview 📚: https://cpython-previews--119674.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

jb2170 committed:
Simplify existing control flow logic of shlex.quote

Removes early return for the special case of an empty string

jb2170 committed:
Add 'always' kwarg to shlex.quote

jb2170 committed:
Update shlex docs for shlex.quote, adding info about the always kwarg

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

vstinner commented on Doc/library/shlex.rst:
```diff
@@ -49,11 +49,13 @@ The :mod:`shlex` module defines the following functions:
    Return a shell-escaped version of the string *s*.  The returned value is a
    string that can safely be used as one token in a shell command line, for
-   cases where you cannot use a list.
+   cases where you cannot use a list. If the *always* keyword argument is
```
Can you please put your sentence in a new paragraph?

jb2170 replied:
Addressed in 5f74da5d9ec1a24b2663364a974e70cb257016ed

vstinner commented on Lib/shlex.py:
```diff
@@ -320,17 +320,14 @@ def join(split_command):
-    # use single quotes, and put single quotes into double quotes
-    # the string $'b is then quoted as '$'"'"'b'
-    return "'" + s.replace("'", "'\"'\"'") + "'"
+    if always or not s or _find_unsafe(s) is not None:
```
I would prefer to keep the special code path for the empty string (`not s`), with just `return "''"`.

jb2170 replied:
For clarity's sake I agree on re-reading.

The reason for the 'simplification' in 5be56a0 was to avoid having a second `"''"` string as well as the `"'" + s.replace("'", "'\"'\"'") + "'"`, a sort of single-source-of-truth / single source of quoting. :smile: 

But really the empty string *still is* a special case (because it requires the `not s` check). All that's happened is the boolean logic on line 325 has been made more complex. Also skipping the `.replace` for the empty string may save a few microseconds lol.

Thus I will revert the special case.

jb2170 replied:
Addressed in 36be111edab30efb36df553e0ab51dba4c8b3615

vstinner commented on Doc/library/shlex.rst:
```diff
@@ -49,11 +49,13 @@ The :mod:`shlex` module defines the following functions:
-   cases where you cannot use a list.
+   cases where you cannot use a list. If the *always* keyword argument is
+   set to ``True`` then the string *s* will always be escaped, even in the
+   absence of special characters.
```
You should explain the purpose of the parameter: it's for performance, no?

jb2170 replied:
Addressed in 5f74da5d9ec1a24b2663364a974e70cb257016ed

vstinner reviewed (commented):
Thanks for your interesting change.

Please add tests to Lib/test/test_shlex.py. Also, you can copy the NEWS entry in Doc/whatsnew/3.14.rst, in a new "shlex" sub-section of Improved Modules.

vstinner commented on Doc/library/shlex.rst:
```diff
@@ -49,11 +49,13 @@ The :mod:`shlex` module defines the following functions:
 
 
-.. function:: quote(s)
+.. function:: quote(s, always=False)
```
To maximize backward compatibility, I suggest to make the parameter keyword-only:

```suggestion
.. function:: quote(s, *, always=False)
```

jb2170 replied:
For forwards compatibility's sake? (In case new kwargs are added to `quote`?). I'm not sure if this is necessary for any backwards compatibility's sake; I've had a look at PEP-3102, but I'd be fine with this addition.

jb2170 replied:
Addressed in 9e9eff0047d731a42f972e5ee20eec6d30503a30

jb2170 committed:
Keep the empty string clearly a special case for shlex.quote

jb2170 committed:
Make `always` a keyword-only argument for shlex.quote

jb2170 committed:
Update shlex docs explaining reasoning for `quote(..., always=True)`

jb2170 committed:
Add testQuoteAlways to shlex tests

The first assertTrue test checks all strings now start with a single quote, ie they have actually been escaped

The second test of assertFalse checks that not all the strings have been escaped when always=False (the default value), since the first, third, and fifth string do not need escaping unless requested.

jb2170 committed:
Add `shlex` subsection to whatsnew for `shlex.quote` additions

vstinner commented on Lib/test/test_shlex.py:
```diff
@@ -337,6 +337,18 @@ def testQuote(self):
+
+        # guarantee escaping all strings
+        strs_always_escaped = [shlex.quote(s, always=True) for s in strs]
+        self.assertTrue(all(s.startswith("'") for s in strs_always_escaped))
```
I would prefer assertEqual() test to test the whole strings, not just the first character. Same below.

jb2170 replied:
Addressed in ffdc6658874efc4fc47d1720d442c5ad08c4cd7b. I've coded in the expected results explicitly; it makes the testing a bit more thorough yes.

jb2170 committed:
Explicitly write the expected results of `test_shlex.testQuoteAlways`

vstinner reviewed (approved):
LGTM

picnixz commented on Misc/NEWS.d/next/Library/2024-05-28-17-24-19.gh-issue-119670.P4J3vX.rst:
```diff
@@ -0,0 +1,2 @@
+:func:`shlex.quote` now has an *always* keyword argument for forcing
+the escaping of the string passed to it
```
```suggestion
Allow :func:`shlex.quote` to unconditionally escape its input
via the keyword-only argument *always*.
```

picnixz commented on Doc/whatsnew/3.14.rst:
```diff
@@ -183,6 +183,12 @@ ___
+-----
+
+* Add keyword-only argument of *always* to :func:`shlex.quote` for forcing
+  the escaping of the string passed to it.
```
```rst
* Allow :func:`shlex.quote` to unconditionally escape its input
  via ``shlex.quote(string, always=True)``.

  (Contributed by YOUR_NAME_OR_GITHUB_NICKNAME in :gh:`119670`.)
```

picnixz commented on Doc/library/shlex.rst:
```diff
@@ -49,12 +49,21 @@ The :mod:`shlex` module defines the following functions:
 
+   If the *always* keyword argument is set to ``True`` then the string *s*
+   will always be escaped, even in the absence of special characters. This is
+   nice for uniformity, for example when escaping a list of strings.
```
```suggestion
   If *always* is ``True``, then the string *s* is
   unconditionally escaped, even in the absence of
   special characters. This is typically useful for
   uniformly escaping a list of strings:
```

picnixz reviewed (commented):
Personally, I'd prefer "force" rather than "always" but that's bike-shedding (using "force" is more common when you want to force a behaviour I think).

jb2170 committed:
Simplify `shlex.quote`'s behaviour to do what it says on the tin
